sklearn/cluster/_hdbscan/_reachability.pyx

```diff
@@ -48,78 +63,125 @@ def mutual_reachability(distance_matrix, min_points=5, max_dist=0.0):
        In Pacific-Asia Conference on Knowledge Discovery and Data Mining
        (pp. 160-172). Springer Berlin Heidelberg.
     """
-    # Account for index offset
-    min_points -= 1
-
-    # Note that in both routines `distance_matrix` is operated on in-place. At
-    # this point, if out-of-place operation is desired then this function
-    # should have been passed a copy.
+    further_neighbor_idx = min_samples - 1
     if issparse(distance_matrix):
-        return _sparse_mutual_reachability(
-            distance_matrix,
-            min_points=min_points,
-            max_dist=max_dist
-        ).tocsr()
+        if distance_matrix.format != "csr":
+            raise ValueError(
+                "Only sparse CSR matrices are supported for `distance_matrix`."
+            )
+        _sparse_mutual_reachability_graph(
+            distance_matrix.data,
+            distance_matrix.indices,
+            distance_matrix.indptr,
+            distance_matrix.shape[0],
+            further_neighbor_idx=further_neighbor_idx,
+            max_distance=max_distance,
+        )
+    else:
+        _dense_mutual_reachability_graph(
+            distance_matrix, further_neighbor_idx=further_neighbor_idx
+        )
+    return distance_matrix
 
-    return _dense_mutual_reachability(distance_matrix, min_points=min_points)
 
-cdef _dense_mutual_reachability(
-    cnp.ndarray[dtype=cnp.float64_t, ndim=2] distance_matrix,
-    cnp.intp_t min_points=5
+def _dense_mutual_reachability_graph(
+    cnp.ndarray[dtype=floating, ndim=2] distance_matrix,
+    cnp.intp_t further_neighbor_idx,
 ):
-    cdef cnp.intp_t i, j, n_samples = distance_matrix.shape[0]
-    cdef cnp.float64_t mr_dist
-    cdef cnp.float64_t[:] core_distances
+    """Dense implementation of mutual reachability graph.
+
+    The computation is done in-place, i.e. the distance matrix is modified
+    directly.
 
-    # Compute the core distances for all samples `x_p` corresponding
-    # to the distance of the k-th farthest neighbours (including
-    # `x_p`).
+    Parameters
+    ----------
+    distance_matrix : ndarray of shape (n_samples, n_samples)
+        Array of distances between samples.
+
+    further_neighbor_idx : int
+        The index of the furthest neighbor to use to define the core distances.
+    """
+    cdef:
+        cnp.intp_t i, j, n_samples = distance_matrix.shape[0]
+        floating mutual_reachibility_distance
+        floating[:] core_distances
+
+    # We assume that the distance matrix is symmetric. We choose to sort every
+    # row to have the same implementation than the sparse case that requires
+    # CSR matrix.
     core_distances = np.partition(
-        distance_matrix,
-        min_points,
-        axis=0,
-    )[min_points]
+        distance_matrix, further_neighbor_idx, axis=1
+    )[:, further_neighbor_idx]
```

Comment:
Using `[:, further_neighbor_idx]` here, means that `core_distances` is not contiguous anymore. Does this lead to a performance regression here?

I'm okay with partitioning in the original code, to keep `core_distances` contiguous.

Also, I think that slicing the original 2d array from `np.partition` is another instance of https://github.com/scikit-learn/scikit-learn/issues/17299. The work around is to use `further_neighbor_idx` to index `core_distances` when computing the `mutual_reachibility_distance`.

Reply:
Turns out there is no significant performance or memory difference between the current non-contiguous setup and one where c-contiguity is enforced:
```python
from sklearn.metrics.pairwise import euclidean_distances
from sklearn.datasets import make_blobs
from sklearn.cluster._hdbscan._reachability import _dense_mutual_reachability_graph

X, _ = make_blobs(n_samples=20_000, random_state=10)
D = euclidean_distances(X)
%timeit -n 5 _dense_mutual_reachability_graph(D, 5)
```

```
non-contig: 2.66 s ± 73.3 ms per loop (mean ± std. dev. of 7 runs, 5 loops each)
contig: 2.74 s ± 197 ms per loop (mean ± std. dev. of 7 runs, 5 loops each)
```
```
non-contiguous / contiguous
Total number of allocations: 92508/93065
Total number of frames seen: 7109/7115
Peak memory usage: 6.6 GB/6.6 GB
```
Currently we do not _require_ that `distance_matrix` is actually contiguous, however that would need to be the case if we enforce that `core_distances` is contiguous. Granted there's no performance or memory gains, I'm not sure it's worth changing.

Reply:
Thanks for the benchmark! Probably, the delta needs to be assessed based on what takes the longer to run on `_dense_mutual_reachability_graph`. This is inspectable using profilers like `py-spy` for instance.

Given the scope of this PR, I am OK pursuing as such and exploring potential performance improvements in subsequent PRs. What do you think?

Reply:
Thanks for the tip to use `py-spy`, it's quite nice for this. 

Looking at it again, it seems that swapping the`prange` to the outer loop and enforcing a contiguous 1D `core_distances` array is significantly faster. Before, the dense loop occupied `~30%` of the runtime of `_dense_mutual_reachability_graph` whereas with the changes the new proportion is `~1%`. I tested whether indexing into the 2D `core_distances` or creating a 1D `core_distances` is preferable, and it seems the latter is notably faster, most likely due to caching.

The new forced contiguity system is still faster even when dealing with non-contiguous inputs.

Reply:
Thanks for the report.

I think posting `.svg` profiles of `py-spy` results interpretable by [Speedscope](https://www.speedscope.app/) with:

```
py-spy record --native -o py-spy-profile.svg -f speedscope -- python ./perf.py
```

might be the most adapted for others to inspect.